File: server/src/core/GameEngine.ts
import { Server } from "socket.io";

export interface Player {
    userId: string;
    uniqueId: string;
    nickname: string;
    profilePictureUrl: string;
    countryCode: string;
    contribution: number;
}

export interface Country {
    score: number;
    contributors: Set<string>;
}

export class GameEngine {
    private io: Server;
    private players: Map<string, Player>;
    private countries: Map<string, Country>;
    private gameActive: boolean;
    private gameTimer: number;
    private timerInterval: NodeJS.Timeout | null;
    private stateUpdatePending: boolean;

    constructor(io: Server) {
        this.io = io;
        this.players = new Map<string, Player>();
        this.countries = new Map<string, Country>();
        this.gameActive = false;
        this.gameTimer = 3600;
        this.timerInterval = null;
        this.stateUpdatePending = false;
    }

    public registerPlayer(data: any, countryInput: string): void {
        if (!this.gameActive) return;

        // Standardization Logic (Moved from index.js)
        const countryMap: { [key: string]: string } = {
            'mx': 'Mexico', 'mexico': 'Mexico', 'mex': 'Mexico', '1': 'Mexico',
            'ar': 'Argentina', 'argentina': 'Argentina', '2': 'Argentina',
            'co': 'Colombia', 'colombia': 'Colombia', '3': 'Colombia',
            'es': 'España', 'espana': 'España', 'España': 'España', '4': 'España', 'spain': 'España',
            'us': 'USA', 'usa': 'USA', 'united states': 'USA', '5': 'USA',
            'pe': 'Peru', 'peru': 'Peru', '6': 'Peru',
            'cl': 'Chile', 'chile': 'Chile', '7': 'Chile',
            'ec': 'Ecuador', 'ecuador': 'Ecuador', '8': 'Ecuador',
            've': 'Venezuela', 'venezuela': 'Venezuela', '9': 'Venezuela',
            'bo': 'Bolivia', 'bolivia': 'Bolivia', '10': 'Bolivia',
            'py': 'Paraguay', 'paraguay': 'Paraguay', '11': 'Paraguay',
            'uy': 'Uruguay', 'uruguay': 'Uruguay', '12': 'Uruguay',
            'sv': 'El Salvador', 'elsalvador': 'El Salvador', '13': 'El Salvador',
            'jp': 'Japon', '14': 'Japon',
            'br': 'Brasil', 'brasil': 'Brasil', '15': 'Brasil',
            'pt': 'Portugal', 'portugal': 'Portugal', '16': 'Portugal',
            'it': 'Italia', 'italia': 'Italia', '17': 'Italia',
            'de': 'Alemania', 'alemania': 'Alemania', '18': 'Alemania',
            'fr': 'Francia', 'francia': 'Francia', '19': 'Francia',
            'gb': 'Reino Unido', 'reino unido': 'Reino Unido', '20': 'Reino Unido',
            'gr': 'Grecia', 'grecia': 'Grecia', '21': 'Grecia',
        };

        const countryCode = countryMap[countryInput.toLowerCase()];
        if (!countryCode) return; // Invalid country

        if (!this.players.has(data.userId)) {
            this.players.set(data.userId, {
                userId: data.userId,
                uniqueId: data.uniqueId,
                nickname: data.nickname,
                profilePictureUrl: data.profilePictureUrl,
                countryCode: countryCode,
                contribution: 0
            });

            if (!this.countries.has(countryCode)) {
                this.countries.set(countryCode, { score: 0, contributors: new Set() });
            }
            this.countries.get(countryCode)?.contributors.add(data.userId);

            // console.log(`User ${data.uniqueId} joined ${country}`);
        }
    }

    public addPoints(userId: string, points: number): void {
        if (!this.gameActive) return;

        const player = this.players.get(userId);
        if (player) {
            player.contribution += points;

            const countryData = this.countries.get(player.countryCode);
            if (countryData) {
                countryData.score += points;
                this.stateUpdatePending = true;
            }

            // Emit immediate like/gift event for animations (bypass batching)
            // Logic handled by caller or we can emit here if purely visual
        }
    }

    public startGame(): void {
        this.gameActive = true;
        this.gameTimer = 3600;
        this.players.clear();
        this.countries.clear();
        this.stateUpdatePending = true;

        if (this.timerInterval) clearInterval(this.timerInterval);

        this.timerInterval = setInterval(() => {
            if (this.gameTimer > 0) {
                this.gameTimer--;
                if (this.gameTimer % 1 === 0) this.stateUpdatePending = true; // Sync timer
                this.broadcastState();
            } else {
                this.finishGame();
            }
        }, 100);
    }

    private broadcastState(): void {
        if (!this.stateUpdatePending) return;

        const countriesArray = Array.from(this.countries.entries()).map(([code, data]) => ({
            code,
            score: data.score,
            contributorCount: data.contributors.size
        })).sort((a, b) => b.score - a.score);

        this.io.emit('gameState', {
            countries: countriesArray,
            timer: this.gameTimer,
            isActive: this.gameActive
        });

        this.stateUpdatePending = false;
    }

    private finishGame(): void {
        this.gameActive = false;
        if (this.timerInterval) clearInterval(this.timerInterval);

        const winners = Array.from(this.countries.entries())
            .map(([code, data]) => ({ code, score: data.score }))
            .sort((a, b) => b.score - a.score);

        const topContributors = Array.from(this.players.values())
            .sort((a, b) => b.contribution - a.contribution)
            .slice(0, 3);

        this.io.emit('gameOver', {
            winners,
            topContributors
        });
    }

    public isActive(): boolean {
        return this.gameActive;
    }

    public getPlayer(userId: string): Player | undefined {
        return this.players.get(userId);
    }
}

File: server-python/game_engine.py
import asyncio
from typing import Dict, List, Optional
from pydantic import BaseModel

class Player(BaseModel):
    userId: str
    uniqueId: str
    nickname: str
    profilePictureUrl: str
    countryCode: str
    contribution: int = 0

class CountryScore(BaseModel):
    code: str
    score: int = 0
    contributorCount: int = 0

class GameEngine:
    def __init__(self, sio):
        self.sio = sio
        self.players: Dict[str, Player] = {}
        self.countries: Dict[str, Dict] = {}
        self.game_active = False
        self.game_timer = 3600
        self.timer_task = None

    def register_player(self, data: any, country: str):
        unique_id = data.get('uniqueId')
        
        country_map = {
            'mx': 'Mexico', 'mexico': 'Mexico', 'mex': 'Mexico', '1': 'Mexico',
            'ar': 'Argentina', 'argentina': 'Argentina', '2': 'Argentina',
            'co': 'Colombia', 'colombia': 'Colombia', '3': 'Colombia',
            'es': 'España', 'espana': 'España', 'españa': 'España', '4': 'España', 'spain': 'España',
            'us': 'USA', 'usa': 'USA', 'united states': 'USA', '5': 'USA',
            'pe': 'Peru', 'peru': 'Peru', '6': 'Peru',
            'cl': 'Chile', 'chile': 'Chile', '7': 'Chile',
            'ec': 'Ecuador', 'ecuador': 'Ecuador', '8': 'Ecuador',
            've': 'Venezuela', 'venezuela': 'Venezuela', '9': 'Venezuela',
            'bo': 'Bolivia', 'bolivia': 'Bolivia', '10': 'Bolivia',
            'py': 'Paraguay', 'paraguay': 'Paraguay', '11': 'Paraguay',
            'uy': 'Uruguay', 'uruguay': 'Uruguay', '12': 'Uruguay',
            'sv': 'El Salvador', 'elsalvador': 'El Salvador', '13': 'El Salvador',
            'jp': 'Japon', '14': 'Japon',
            'br': 'Brasil', 'brasil': 'Brasil', '15': 'Brasil',
            'pt': 'Portugal', 'portugal': 'Portugal', '16': 'Portugal',
            'it': 'Italia', 'italia': 'Italia', '17': 'Italia',
            'de': 'Alemania', 'alemania': 'Alemania', '18': 'Alemania',
            'fr': 'Francia', 'francia': 'Francia', '19': 'Francia',
            'gb': 'Reino Unido', 'reino unido': 'Reino Unido', '20': 'Reino Unido',
            'gr': 'Grecia', 'grecia': 'Grecia', '21': 'Grecia',
        }
        
        country_code = country_map.get(country.lower().strip())
        if not country_code:
            return  # Invalid country or random comment
            
        if unique_id and unique_id not in self.players:
            self.players[unique_id] = Player(
                userId=data.get('userId', ''),
                uniqueId=unique_id,
                nickname=data.get('nickname'),
                profilePictureUrl=data.get('profilePictureUrl'),
                countryCode=country_code
            )
            
            # Ensure country exists in scores
            if country_code not in self.countries:
                self.countries[country_code] = {"score": 0, "contributors": set()}
            
            self.countries[country_code]["contributors"].add(unique_id)

    def add_points(self, unique_id: str, points: int):
        player = self.players.get(unique_id)
        if player:
            player.contribution += points
            code = player.countryCode
            if code in self.countries:
                self.countries[code]["score"] += points
                # Re-add just in case
                self.countries[code]["contributors"].add(unique_id)
            
            asyncio.create_task(self.emit_game_state())

    async def start_game(self):
        self.game_active = True
        self.game_timer = 3600 # Reset timer
        if self.timer_task:
            self.timer_task.cancel()
        self.timer_task = asyncio.create_task(self.tick())

    async def stop_game(self):
        self.game_active = False
        if self.timer_task:
            self.timer_task.cancel()
            self.timer_task = None
        await self.emit_game_state()

    async def tick(self):
        while self.game_active and self.game_timer > 0:
            await asyncio.sleep(1)
            self.game_timer -= 1
            await self.emit_game_state()
            if self.game_timer <= 0:
                await self.stop_game()

    def get_game_state(self):
        sorted_countries = sorted(
            [
                {
                    "code": code,
                    "score": data["score"],
                    "contributorCount": len(data["contributors"])
                }
                for code, data in self.countries.items()
            ],
            key=lambda x: x["score"],
            reverse=True
        )
        return {
            "countries": sorted_countries,
            "timer": self.game_timer,
            "isActive": self.game_active
        }

    async def emit_game_state(self):
        state = self.get_game_state()
        await self.sio.emit('gameState', state)

    def is_active(self):
        return self.game_active

    def get_player(self, unique_id: str):
        return self.players.get(unique_id)
